### ISSUE TITLE
Middleware must create objects before attempting dispose even if object never needed during request

### ISSUE THREAD

**Issue by mr-robert** [NathanLBCooper/OwinPerRequestUnityMiddleware#2]
Not entirely sure this falls under the realm of issue if it can't be fixed, or how it could be fixed.

The issue I see is that inside the RequestLifetimeMiddleware.cs  Invoke method, all registrations are retrieved. For each registration, Resolve is called, which in turn creates instances of all PerRequestLifetime objects, even ones that were never used in the original request.

For example, if you have a web api project with two controllers, and each controller has one unique class that is in PerRequestLifetime, after a call to either controller, the controller makes one of the objects, the request ends, and while Invoke is cleaning up, the other object gets created, just to get disposed on the very next line.

It seems that there should really be a way of not "Resolving" objects that were not previously resolved during the request. If there is some way of telling that an object has been previously resolved in the current request, and just disposing those objects, rather than causing the creation of all unused objects just to dispose them.

I imagine in a monolith project (not that I'm encouraging those) this would be an even greater issue.

What do you think?

**Comment by NathanLBCooper** [comment]
I agree 100%, the whole thing doesn't really work as it currently stands. I suspect that I'd need to dig into Unity a lot more to acheieve this, and probably change the structure of my approach.

I self assigned this a couple of months ago, hoping I'd get round to it, but that hasn't really happened yet. Tbh, this project isn't a priority for me at the moment.

**Comment by codesocket** [comment]
You may have a concurrency issue here. How do you ensure that you don't end up disposing instances of PerRequestLifetimeManager that don't belong to current request?

**Comment by icnocop** [comment]
See [the sample code in the Stack Overflow question](https://stackoverflow.com/q/56844803/90287):

```
    public class PerRequestMiddleware : OwinMiddleware
    {
        [ThreadStatic] static Dictionary<Type, object> _instances;

        public PerRequestMiddleware(OwinMiddleware next) : base(next)
        {
        }

        public override async Task Invoke(IOwinContext context)
        {
            _instances = null;

            await Next.Invoke(context);

            if (_instances != null)
            {
                foreach (var disposable in _instances.Values.OfType<IDisposable>())
                {
                    disposable.Dispose();
                }
            }
        }

        public static T GetInstance<T>(Func<T> constructor)
        {
            if (_instances == null)
            {
                _instances = new Dictionary<Type, object>();
            }

            var type = typeof(T);

            if (!_instances.TryGetValue(type, out var instance))
            {
                instance = constructor();
                _instances[type] = instance;
            }

            return (T)instance;
        }
    }
```

Usage:

1. `app.UsePerRequestMiddleware(container);`
2. `container.RegisterFactory<ITestService>(c => PerRequestMiddleware.GetInstance(() => new TestService()));`
